### ISSUE TITLE
Mark F5 modules stableinterface

### ISSUE THREAD

**Issue by caphrim007** [ansible/ansible#40913]
##### SUMMARY

Mark the F5 community modules as `stableinterface` in the module's ANSIBLE_METADATA.

A customer of F5's is interested in using the F5 modules, however their company policy does not allow the use of `preview` modules.

The customer is asking that `preview` status be removed.

The full request is,

"""
Our automation tool of choice is Ansible Tower and while we were happy to see there are F5 modules in v2.5, the bitter side is that all of those modules are flagged as preview.
We would like to use those modules in our automation tasks, but our company policy does not allow the use of preview modules, which is understandable I think.

So my question would be, can we have an estimation when the preview flag is expected to be removed? If that's in the near future, we can start developing our automated tasks in our lab and migrate them to production when the flag is removed. But if there's much to wait, we have to abandon the modules to be able to meet the deadlines.
"""

Requesting discussion of this with the Ansible Core Team as mentioned in documentation here

https://docs.ansible.com/ansible/latest/dev_guide/developing_modules_documenting.html?highlight=ansible_metadata#ansible-metadata-block

##### ISSUE TYPE

 - Feature Idea

##### COMPONENT NAME

all bigip_* and bigiq_* modules

##### ANSIBLE VERSION

```
ansible 2.6.0
  config file = /here/test/integration/ansible.cfg
  configured module search path = ['/here/library/modules']
  ansible python module location = /usr/local/lib/python3.6/site-packages/ansible
  executable location = /usr/local/bin/ansible
  python version = 3.6.5 (default, May  5 2018, 03:09:35) [GCC 4.9.2]
```

##### STEPS TO REPRODUCE

```yaml

```

##### ACTUAL RESULTS

```

```

**Comment by caphrim007** [comment]
@bcoca @sivel @gundalow @Qalthos @privateip requesting your thoughts on this

**Comment by caphrim007** [comment]
With my maintainer hat on, I should outline the current "situation" that F5 is in.

The F5 modules are still under development. While the existing set of params and functionality is more-or-less fixed, there is the continual addition of new features to the existing modules.

F5 has also enjoyed the non-committal flexibility of being in `preview` status. F5 does not support the F5 modules as part of a company wide effort. The development team for the modules, is 1 person. That one person maintains the existing 95 core modules, plus N more modules that are external to core and waiting on feedback before upstreaming.

In some sense, committing to a stable interface would be putting an extra burden on said developer that they do not have today. Note that said developer does more than just write the code; they are also responsible for all existing support, outreach, speaking and customer engagement, training, testing (and infrastructure for this) and travel that is part of this effort.

This should be taken into consideration when adding more to their **required** responsibilities. Said developer, to the best of their abilities attempts to provide a stable interface today, though there is no requirement to do so.

Some of the above may become "better" moving forward, but F5 is unable to commit to that right now.

Maintainer notes that the Networking Team of Ansible is moving and changing much of the code-base. In specifically the area of networking, major functionality has changed in every previous version of Ansible that would have deprecated functionality or changed requirements.

Maintainer is not comfortable with, or capable of, staying up-to-date with all of these changes (and supporting the users through their transitions of these changes) and is deliberately waiting for stable interfaces and more mature code before adapting all F5 integrations to meet it.

At times this has led to the F5 modules appearing "out of date" in terms of how they connect to Ansible (action/terminal plugins) or how they participate in newly minted Networking Team conventions (provider, platform options https://docs.ansible.com/ansible/devel/network/user_guide/platform_index.html).

Maintainer wishes to stress that this is not because they do not care. They care very much. But there is a very real cost to maintainer to have to make changes to stay up to date with this movement.

Soliciting Core Team's feedback on this proposal. Alternatively, is there a way to satisfy the customer request in some other way?

Maintainer wants to see the customer be whole. If this means becoming `stableinterface`, then maintainer is ready to accept that, provided the above concerns are known in advance.

**Comment by eoprede** [comment]
I can chime in from the user's perspective. I am working right now through automating setup of the F5 with this ansible module. I am mostly done going through the bigip.conf parameters and so far I had to:
1) Submit bug fixes for bigip_monitor_http and bigip_monitor_https
2) Add modules bigip_monitor_dns, bigip_profile_fastl4 and bigip_profile_server_ssl
3) Add additional configurable parameters to modules bigip_profile_client_ssl and bigip_profile_tcp

I am sure that I would have to add new modules, for example for packet filters. Unfortunately I do not have sufficient time to properly document them and submit them to Ansible, not to mention that the right way to fix all the modules IMO is to support all of the parameters for given API endpoint, rather than cherry-picking some more common ones.

So basically what I am trying to say - these modules are quite incomplete at the time and need noticeable amount of effort to be production-ready. Most of the effort is not complicated from developmental perspective, but time consuming from documentation stand point. 
I am honestly not sure if given F5 customer would actually be able to automate significant amount of their work with F5s, without putting development effort into modules, which would probably violate their policies as well.

**Comment by caphrim007** [comment]
@eoprede have you raised all of these issues and enhancement requests to F5? Specifically the f5-ansible repository that they maintain.

Have you contacted myself at my email address in my profile to make these things known?

I can say with confidence that F5 will **never** support all our product APIs in Ansible. It's not possible in their current state. The path we've been going down is working with customers to address their automation needs for what they **do** automate, instead of starting at A and going to Z.

There are more than 3000 apis in just 1 of the products, and these can change across versions, so the total number of combinations to develop and test is not possible for the current maintainer to handle.

Email me at my address in my GitHub profile and I'd be happy to escalate your needs up the priority list

**Comment by eoprede** [comment]
@caphrim007 
You can write a module to support arbitrary parameters (you can see my ansible module for Fortigates as an example), but it goes against fixed argument spec policy with Ansible. Maybe F5 could help to change that?

As for contacting F5 - it takes me about 30 minutes to add modifications to or quickly build a new module. Going thorough F5 support (even the premium one) is just not feasible, as it would take days or months to get it done.

**Comment by caphrim007** [comment]
@eoprede I'm not asking you to contact support. I'm asking you to email me. I write all these modules and the Ansible work at F5 does not go through standard support channels. If you want your bugs and features fixed, it is in your best interest to contact me, and not do it yourself.

**Comment by caphrim007** [comment]
I've also read the reasoning for fixed arg spec, and i agree with it. arbitrary params makes docs untenable.

**Comment by eoprede** [comment]
I absolutely understand the reasoning for fixed arg spec, but I also think that it can be extremely limiting - just as we see it here. 

As for contacting you - thanks for the offer. Once I am done with my work, I'll let you know what I needed to change.

**Comment by bcoca** [comment]
approved in linked irc meeting

**Comment by delgadillo22** [comment]
Note that since this was Issue was posted and discussed, F5 announced formal support for our F5 Ansible modules.  Refer to our support policies at https://f5.com/support/support-policies. We continue to invest in Ansible as it provides great value to our customers.
